common/lib/xmodule/xmodule/seq_module.py

```diff
@@ -279,6 +279,74 @@ def verify_current_content_visibility(cls, date, hide_after_date):
             datetime.now(UTC) < date
         )
 
+    def _has_access_check(self, *args):
+        """
+        Helper method created for the following method
+        gate_sequence_if_it_is_a_timed_exam_and_contains_content_type_gated_problems
+        Created to simplify the relevant unit tests
+        This way the has_access call can be patched without impacting other has_access calls made during the test
+        """
+        # importing here to avoid a circular import
+        from lms.djangoapps.courseware.access import has_access
+        return has_access(*args)
+
+    def gate_sequence_if_it_is_a_timed_exam_and_contains_content_type_gated_problems(self):
```

Comment:
🎱 What do you think of structuring this as a function that returns a boolean and doing the actual gating in a separate (or the existing `student_view`) function?

Reply:
i'd like to keep the relevant code contained within this function so we're not cluttering up the calling view with too much of this very specific code . 
we could break things up within this function itself into multiple functions but imo that wouldn't make things clearer since it is a pretty short function as is

Reply:
:8ball: Sure, just thought that breaking up the code into business functionality + applying it pieces would make it easier to test and for someone else to understand, as the business logic appears to be pretty complex and hard to read/evaluate to find the edge cases.

Something like this:
```
def should_gate_timed_exams_with_gated_problems():
  # returns a boolean at the end of the if/else branches

def gate_timed_exams_with_gated_problems(gate_or_not_gate_boolean):
  if true:
     self.gated_sequence_fragment = access.user_fragment
  else:
     self.gated_sequence_fragment = None

def student_view():
 gate_exams = should_gate_timed_exams_with_gated_problems()
 gate_timed_exams_with_gated_problems(gate_exams)
  ..
```

Reply:
i would prefer the call within student_view be limited to a single function